state/machine_linklayerdevices.go

```diff
@@ -471,6 +496,49 @@ func (m *Machine) setDevicesFromDocsOps(newDocs []linkLayerDeviceDoc) ([]txn.Op,
 	return ops, nil
 }
 
+// removeUnusedLinkLayerDevicesOps removes link-layer devices that
+// are not in the in-use list.
+func (m *Machine) removeUnusedLinkLayerDevicesOps(inUse []LinkLayerDeviceArgs) ([]txn.Op, error) {
+	devices, closer := m.st.db().GetCollection(linkLayerDevicesC)
+	defer closer()
+
+	existingToDelete := make(map[string]linkLayerDeviceDoc)
+	filter := bson.D{{
+		"_id", bson.D{{"$regex", "^" + linkLayerDeviceGlobalKey(m.Id(), "")}},
+	}}
+	var docs []linkLayerDeviceDoc
+	err := devices.Find(filter).All(&docs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, d := range docs {
+		existingToDelete[d.Name] = d
+	}
+
+	var ops []txn.Op
+	for _, arg := range inUse {
+		delete(existingToDelete, arg.Name)
+	}
+	for _, devDoc := range existingToDelete {
+		dev := LinkLayerDevice{st: m.st, doc: devDoc}
+		removeOps, err := removeLinkLayerDeviceOps(m.st, dev.DocID(), dev.parentDocID())
+		// TODO(wallyworld) - do we need to remove parent devices also?
```

Comment:
need to confirm if this is needed, and make adjustment to the code as necessary

Reply:
The problem is that this will fail for any parents with children, even if those children are being deleted in the same transaction.

`SetParentLinkLayerDevicesBeforeTheirChildren` handles this in reverse, by eschewing a transaction in favour of batching the update transactions as children-first. It's terrible.

I will see what I can do with it.

Reply:
Merged #11685 to fix this